R/model.R

```diff
@@ -367,19 +374,67 @@ sample_method <- function(data = NULL,
   )
 
   runset <- RunSet$new(args = cmdstan_args, num_runs = num_chains)
-  for (chain in chain_ids) { # FIXME: allow parallelization
-    run_log <- processx::run(
+  procs <- list()
+  cat("Running MCMC with", num_chains, "chain(s) ...\n")
+  for (chain_id in runset$run_ids()) {
+    procs[[chain_id]] <- processx::process$new(
       command = runset$command(),
-      args = runset$command_args()[[chain]],
+      args = runset$command_args()[[chain_id]],
       wd = dirname(self$exe_file()),
-      echo_cmd = FALSE,
-      echo = TRUE
+      echo_cmd = TRUE,
+      stdout = "|",
+      stderr = "|"
     )
+    runset$mark_chain_start(chain_id)
+  }
+  while (any_chain_alive(procs, runset)) {
+    processx::poll(procs, 100)
+    for (chain_id in runset$run_ids()) {
+      output <- procs[[chain_id]]$read_output_lines()
+      runset$process_sample_output(output, chain_id)
+    }
+  }
+  if (num_chains > 1) {
+    num_failed_chains <- num_chains - sum(runset$chain_state() == 5)
+    if (num_failed_chains == 0) {
+        if (num_chains == 2) {
+          cat("Both chains finished succesfully.\n")
+        } else {
+          cat(paste0("All ", num_chains," chains finished succesfully.\n"))
+        }
+        cat(paste0("Mean execution time: ",
+                   format(round(mean(runset$time()$total_time),1), nsmall = 1),
+                   " seconds"))
+    } else {
+      if (num_failed_chains == num_chains) {
+        cat("All chains finished unexpectedly!\n")
+      } else {
+        cat(num_failed_chains, "chain(s) finished unexpectedly!\n")
+        cat("The remaining chains had a mean execution time of",
+            format(round(mean(runset$time()$total_time), 1), nsmall = 1),
+            "seconds")
+      }
+    }
   }
   CmdStanMCMC$new(runset) # see fit.R
 }
 CmdStanModel$set("public", name = "sample", value = sample_method)
 
+any_chain_alive <-function(procs, runset) {
```

Comment:
what do you think about making `$any_chain_alive(procs)` a method for the runset object? not a big deal either way, but seems more consistent to include it in runset with the other methods you added (e.g., `chain_state`). 

If you want to do this then I think you just need to change the while loop to 
```r
while(runset$any_chain_alive(procs))
```
and use `self` instead of `runset` inside the function.

Reply:
Great idea. Will fix.

Reply:
I’m not sure but I wonder if `procs` itself should also go in `runset`. What do you think?

Reply:
Good call. Will finish up this PR tomorrow. Spending my Mondays and Tuesdays TAing till February which makes me less productive :)